[1단계 - 자동차 경주 구현] 이상(이건상) 미션 제출합니다.

kunsanglee opened the pull request:
안녕하세요 알렉스 6기 이상(이건상)입니다 반갑습니다! 서툴지만 페어인 아톰과 열심히 미션을 수행했습니다.
미션을 수행하면서 클래스 관계를 최대한 단순히 구성해 복잡도를 낮추고 페어 프로그래밍을 수행했습니다.

자동차의 위치와 이름을 나타내는 Position, Name 클래스에 getter 메서드를 열어서 출력에 필요한 객체인 CarInfo를 생성할 때
파라미터로 넘겨주었는데 좋은 코드인가에 대해 의문이 있습니다.
또 가능한 모든 클래스를 불변으로 만드는 것이 절대적으로 좋은지에 대해서도 궁금합니다.

le2sky committed:
docs : 기능 목록 작성

le2sky committed:
docs : 누락된 요구사항 추가 및 도메인 용어 정의

le2sky committed:
docs : 기능 목록 수정

이름을 입력 받는 기능과 이름을 생성하는 기능을 분리한다.

le2sky committed:
chore : 불필요한 파일 제거

le2sky committed:
feat : 이름 생성 기능 구현

le2sky committed:
docs : 구현 기능 체크

le2sky committed:
feat : 위치 생성 기능 구현

le2sky committed:
feat : 자동차 생성 기능 구현

le2sky committed:
feat : 라운드 생성 기능 구현

le2sky committed:
docs : 구현 기능 체크

le2sky committed:
docs : 기능 목록 수정

누락된 기능 추가 작성 및 구현 기능 체크

le2sky committed:
feat : 자동차 위치 증가 기능 구현

le2sky committed:
docs : 구현 기능 체크

le2sky committed:
feat : 자동차 전진 기능 구현

le2sky committed:
feat : 경주 기능 구현

le2sky committed:
docs : 기능 목록 수정

누락된 기능 추가 작성 및 구현 기능 체크

le2sky committed:
docs : 기능 목록 수정

- 라운드 수 만큼 경주를 진행할 수 있다는 요구사항이 누락되어 추가했다.
- 라운드 수 만큼 경주를 진행하는 기능을 구현해서 체크했다.

le2sky committed:
feat : 라운드 수 만큼 경주를 진행하는 기능 구현

le2sky committed:
docs : 기능 목록 수정

- 기능 목록을 정렬했다.
- 구현 기능을 체크했다.

le2sky committed:
feat : 입출력 기능 구현

- 라운드 수 입력 기능을 구현했다.
- 이름 입력 기능을 구현했다.
- 라운드 결과 출력 기능을 구현했다.

le2sky committed:
docs : 우승자 출력 기능 구현 체크

le2sky committed:
feat : 우승자 출력 기능 구현

le2sky committed:
docs : mermaid 그래프 추가

le2sky committed:
refactor : 패키지 정리

le2sky committed:
refactor : 도메인 코드 개선

le2sky committed:
refactor : 입출력 기능 가독성 개선

le2sky committed:
refactor : 테스트 코드 가독성 개선

yxxnghwan commented on src/main/java/racinggame/Main.java:
```diff
@@ -0,0 +1,22 @@
+import racinggame.domain.Round;
+import racinggame.domain.condition.RandomMoveCondition;
+
+public class Main {
```
![스크린샷 2024-02-16 오전 3 34 11](https://github.com/woowacourse/java-racingcar/assets/63737500/48796036-f466-414c-ab07-994402af5b01)

👀

kunsanglee replied:
자동차들을 관리하는 일급 컬렉션 Cars 클래스를 만들어서 내부적으로 검증 후 에러를 반환하도록 수정하였습니다.

yxxnghwan commented on src/main/java/racinggame/Main.java:
```diff
@@ -0,0 +1,22 @@
+public class Main {
+
+    public static void main(String[] args) throws IOException {
+        List<Car> cars = InputView.readNames();
```
메소드 이름을 보면 이름들을 반환할 것 같은데 List<Car>를 반환하네요!

kunsanglee replied:
놓치고 지나간 부분인데 봐주셔서 감사합니다! InputView 클래스의 다른 메서드인 readRound() 메서드 경우에는 Round 객체를 생성하여 반환하는 것에 통일하여 readNames 메서드의 이름을 readCars로 변경하여 Cars 객체를 반환하도록 변경했습니다.

yxxnghwan commented on src/main/java/racinggame/domain/Car.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public static Car from(String name) {
+        return new Car(Name.from(name), Position.from(0));
```
Position의 경우 0으로 초기화 되는데, 정적 팩토리 메소드를 쓰는 김에 장점을 좀 더 살려보면 어떨까요?
정적 팩토리 메소드는 이름을 지어줄 수 있습니다!

kunsanglee replied:
단순히 매개변수 1개를 받으니까 from이라는 메서드명을 부여했었는데, 생각해보니 기본 위치값을 지닌 자동차라는 의미 전달이 잘 되지 않았던 것 같습니다. 그래서 createDefaultPositionCar 라는 메서드 이름으로 수정하여 반영했습니다!

yxxnghwan commented on src/main/java/racinggame/domain/MoveCondition.java:
```diff
@@ -0,0 +1,6 @@
+package racinggame.domain;
+
+public interface MoveCondition {
+
+    boolean isMovable();
+}
```
이 인터페이스를 구현하는 구현체는 하나인데, 왜 클래스가 아니라 인터페이스로 구현하셨나요? 🤔

kunsanglee replied:
MoveCondition 인터페이스의 메서드로 boolean 타입을 반환하는 추상메서드 isMovable이 있는데,
랜덤에 대한 테스트를 용이하게 하기 위해 따로 클래스까지 작성해야 하나? 라는 생각으로 MoveCondition이 필요한 테스트 진행시 람다로 넘겨서 단순히 true를 반환하도록 처리하였으나,

RandomMoveCondition 클래스를 만들었으면 테스트도 테스트용 클래스를 만들어 통일하거나
런타임에 사용되는 코드와 테스트코드 양쪽에 모두 람다를 사용하거나 통일을 했어야 적절했다고 생각합니다.
그래서 리팩토링 진행은 테스트용 구현체인 FixedMoveCondition을 작성하여 반영했습니다.

혹시 알렉스가 생각하기에 두 방식 중 어떤 쪽을 더 선호하시고 실무에서 사용하시는지 의견이 궁금합니다!

yxxnghwan replied:
인터페이스로 추상화해주신 덕분에 테스트하기가 수월해져서 좋은 구조라고 생각합니다.😁

람다로 쓰는 것과 클래스를 만들어서 사용하는 것에 대한 저의 생각을 정리하자면
클래스를 만들어 줄 경우 해당 클래스에 이름을 붙여 어떤 동작 및 의도를 가지는지 표현해줄 수 있어 좋은 것 같고, 람다의 경우 사용할 때 좀 더 편하게 쓴다는 장점이 있을 것 같아요. 정답은 없으니 이상이 생각하는 장단점을 가지고 사용의 기준을 정해보면 좋을 것 같습니다!

그리고 익명 클래스를 인스턴스화 할 때 람다로 쓰는 것과 `new` 키워드를 붙여 인스턴스화 하는 것에 대한 차이도 고민해보면 좋을 것 같아요!

yxxnghwan commented on src/main/java/racinggame/domain/Name.java:
```diff
@@ -0,0 +1,48 @@
+
+    private static void checkNameIsEmpty(String name) {
+        if (name.isEmpty()) {
+            throw new RuntimeException();
```
좀더 친절한 에러를 던져주면 좋을 것 같아요. 어떤 케이스로 인해 에러가 발생했는지 알면 좋지 않을까요?

kunsanglee replied:
예외 상황에 맞게 구체적 예외를 던지도록 수정하여 반영했습니다.

yxxnghwan commented on src/main/java/racinggame/domain/Name.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    private static void checkNameLength(String name) {
+        if (name.length() > 5) {
```
매직넘버는 상수화하면 이름을 지어주고 값의 의미를 나타낼 수 있습니다!

kunsanglee replied:
네오의 강의 중 매직넘버를 보고 의미 전달이 잘 되면 굳이 상수로 선언하지 않아도 된다 라는 뉘앙스로 들어서
페어 프로그래밍시 약간은 일부러(?) 상수 선언하지 말아보자! 하고 진행했었습니다.

알렉스가 실무에서 코드를 작성할 때 모든 매직넘버를 상수화하는지, 아니면 섞어서 하는지 궁금합니다!

yxxnghwan replied:
상수로 사용했을 경우 장점은 의미 전달을 쉽게 하는 것도 있지만, 재사용의 장점도 가져갑니다! 
해당 의미의 값을 사용하는 곳이 여러 곳이 된다면 모두 찾아서 수정해주어야해서 유지보수성이 떨어질 것 같아요.

당장 에러 메시지 안에서도 5라는 숫자를 가지고 있고, 요구사항 변경시 그 쪽에서도 같이 변경해줘야할 것 같네요!

```
private static void checkNameLength(String name) {
        if (name.length() > MAX_NAME_LENGTH) {
            throw new IllegalArgumentException(String.format("자동차 이름은 최대 5자 이하만 가능합니다. 입력한 자동차 이름: %s", name));
        }
    }
```

yxxnghwan commented on src/test/java/racinggame/domain/NameTest.java:
```diff
@@ -0,0 +1,33 @@
+    @Test
+    void checkNameLength() {
+        assertThatThrownBy(() -> Name.from("abcdef"))
+            .isInstanceOf(RuntimeException.class);
+    }
+}
```
파일 끝에 개행문자를 추가하는걸 추천드립니다. EOF에 대해서 알아보아요

yxxnghwan commented on src/main/java/racinggame/domain/RacingGame.java:
```diff
@@ -0,0 +1,84 @@
+
+    public List<RoundResult> getResult() {
+        return results;
+    }
```
꼭 필요한 메소드일까요? race() 메소드의 결과로 응답해주는건 어떨까요?

kunsanglee replied:
기존에 경주의 결과를 RacingGame 내부에 저장해두고 추후 결과의 다양한 처리(5라운드 진행하고 3라운드에 가장 멀리 갔던 자동차를 뽑는다던가)에 대응할 수 있도록 만들려고 생각했었습니다. 그래서 커맨드와 쿼리를 분리하여 race 메서드는 경주를 진행하는 커맨드만 수행하고, getResult와 같은 메서드가 쿼리를 담당하도록 작성했었는데, 객체지향 메세지 관점에서 적절치 않았던 것인지 고민입니다.

현재는 피드백 반영하여 내부에 List<RoundResult>를 상태로 가지지 않고 race 메서드 내부에서 경주 결과를 집계하여 반환하도록 수정하였습니다.

알렉스가 보시기에는 어떤가요?

yxxnghwan replied:
조회를 분리하는 확장성을 고려하셨군요! 개인적으론 결과를 저장하고 관리하는 역할의 클래스가 따로 필요할 것 같은데 해당 부분에 대한 고민은 이번 미션에선 많이 앞서나간 고민으로 보여서 앞으로 학습하며 이상만의 생각을 완성시켜나가면 좋을 것 같습니다!

race() 메소드의 응답으로 경주의 결과를 반환하는 것은 자연스럽게 느껴집니다. 커맨드라고 해서 응답이 없어야한다는 룰은 없으니까요!

많은 고민을 하고 미션 수행하고 계신 것 같아 좋네요!👍🏿

yxxnghwan commented on src/main/java/racinggame/domain/condition/RandomMoveCondition.java:
```diff
@@ -0,0 +1,14 @@
+
+public class RandomMoveCondition implements MoveCondition {
+
+    private static final Random random = new Random();
```
static final인데 네이밍이 소문자인 이유는 무엇일까요?
상수와는 어떤 점이 다른가요?

yxxnghwan commented on docs/README.md:
```diff
@@ -0,0 +1,44 @@
+## 도메인 용어 정리
+
```
문서 작성 👍🏿

yxxnghwan commented on src/main/java/racinggame/InputView.java:
```diff
@@ -0,0 +1,36 @@
+    private InputView() {
+    }
+
+    public static List<Car> readNames() throws IOException {
```
IOException을 외부로 던지네요! 해당 클래스가 IO 작업을 담당하는 역할을 하는 것 같은데 이 클래스 안에서 해결할 수는 없을까요?

kunsanglee replied:
BufferedReader를 사용하는 메서드 readLine에서 try-catch를 사용하여 에러를 핸들링하도록 반영했습니다.
감사합니다 알렉스!!!

yxxnghwan reviewed (changes requested):
안녕하세요 이상, 이번 미션 리뷰하게 된 알렉스입니다.
미션 잘 구현해주셨네요.
몇몇 코멘트 남겼으니 확인 부탁드려요.
이해가 어려운 리뷰가 있다면 DM 남겨주셔도 좋습니다😁

> 자동차의 위치와 이름을 나타내는 Position, Name 클래스에 getter 메서드를 열어서 출력에 필요한 객체인 CarInfo를 생성할 때
> 파라미터로 넘겨주었는데 좋은 코드인가에 대해 의문이 있습니다.
> 또 가능한 모든 클래스를 불변으로 만드는 것이 절대적으로 좋은지에 대해서도 궁금합니다.
위 질문에 대한 저의 생각은 getter의 경우 필요하다면 사용해도 좋다고 봅니다. 단 getter도 객체의 상태를 외부에 노출시키기 때문에 상태를 변경할 여지를 줄 수 있으니 고민해보시면 좋을 것 같아요. 불변으로 설계된 클래스라면 안전하겠네요!

클래스가 불변으로 설계된다면 예상치 못한 곳에서 상태를 변경하여 발생하는 버그를 방지할 수 있습니다. 다만, 상태가 바뀌는 것이 자연스러운 클래스가 존재할 수 있다고 생각하므로 항상 불변으로 설계하는 것이 옳은 건 아닌 것 같아요. 어떤 경우에 불변이 좋고 어떤 경우에 불변이 아니어도 될지 이상만의 기준을 세워보면 좋을 것 같네요😁

kunsanglee committed:
style : code style 적용 및 EOF 추가

kunsanglee committed:
refactor : 구체적인 예외를 던지도록 수정

kunsanglee committed:
refactor : 상수 변수명을 대문자로 수정

kunsanglee committed:
refactor : 매직넘버 및 하드코딩된 문자 상수로 선언

kunsanglee committed:
refactor : 자동차들을 관리하는 일급 컬렉션 Cars를 사용하도록 변경

kunsanglee committed:
refactor : 기본 위치값을 가진 자동차를 생성하는 팩토리 메서드 수정

- 기존에 매개변수를 한 개만 받아서 기본값을 할당해서 자동차를 반환하는 정적 팩토리 메서드였는데
메서드의 네이밍을 기본값을 가진 자동차를 생성한다는 의미로 변경하였다.
- 자동차의 위치를 나타내는 Position 클래스에 기본 위치를 가지는 Position 객체를 초기화해두고 사용하도록 변경하였다.

kunsanglee committed:
test : Cars 테스트코드 작성 및 RacingGameTest에 작성된 테스트 삭제

kunsanglee committed:
refactor : RacingGame 클래스가 경주의 결과를 반환하도록 수정

- RacingGame의 race 메서드를 호출하면 인스턴스 변수로 가지고있던 List<RoundResult> results에 라운드별 결과를 저장해두었는데,
race 메서드 호출시 지역변수로 결과 리스트를 생성해서 바로 반환하는 로직으로 수정했다.
기존의 생각으로는 결과를 집계해서 가지고있다가 결과를 가공하는 상황, 예를들어
5회 레이싱 결과를 마치고 3회까지 가장 긴 거리를 움직인 자동차를 반환한다던지 하는 특수한 상황에 대비하여
내부적으로 저장하는 방식을 선택하여 race 메서드를 커맨드와 쿼리를 분리하여 getCars, getResult 와 같은
결과를 반환하는 쿼리성 메서드를 분리하였었다.

kunsanglee committed:
refactor : InputView의 메서드에서 에러를 핸들링하도록 수정

kunsanglee committed:
refactor : MoveCondition의 테스트용 구현체 FixedMoveCondition 작성 및 테스트코드 수정

- MoveCondition 인터페이스의 구현체인 RandomMoveCondition 클래스는 실제 사용자들이 게임을 진행할 때
랜덤 로직을 수행하도록 하지만, 테스트시 모든 자동차들이 움직일 수 있도록 간단하게 true를 반환하는
테스트 목적 구현체를 따로 작성하지 않고 람다식을 매개변수로 넣어서 처리했었다.

코드의 가독성을 위해 프로덕션 코드에서도 람다를 사용하거나, 테스트용 구현체를 명시적으로 작성하도록
통일하는 것이 더 좋을 것 같아 테스트용 구현체 FixedMoveCondition 클래스를 작성하여 테스트코드를 수정했다.

kunsanglee committed:
refactor : InputView에서 자동차 이름을 입력받을 때 공백을 제거하도록 수정

yxxnghwan commented on src/main/java/racinggame/domain/Name.java:
```diff
@@ -0,0 +1,54 @@
+
+    private static void checkNameLength(String name) {
+        if (name.length() > MAX_NAME_LENGTH) {
+            throw new IllegalArgumentException(String.format("자동차 이름은 최대 5자 이하만 가능합니다. 입력한 자동차 이름: %s", name));
```
MAX_NAME_LENGTH가 수정되면 해당 에러메시지도 수정되어야 하겠네요!

kunsanglee replied:
아..! 정말 날카로우시군요 알렉스 👍 수정하여 2단계 미션에서 반영하겠습니다!!

yxxnghwan commented on src/main/java/racinggame/domain/Position.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    private static void checkIsNonNegative(int position) {
+        if (position < ZERO) {
+            throw new RuntimeException();
+        }
```
여기서도 구체적인 에러를 던져주면 좋을 것 같습니다!

kunsanglee replied:
더 철저히 코드를 돌아보고 PR 하도록 노력하겠습니다.. 죄송합니다!

yxxnghwan commented on src/main/java/racinggame/domain/Position.java:
```diff
@@ -0,0 +1,53 @@
+
+    private static final int ZERO = 0;
+    private static final int INCREASE_AMOUNT = 1;
+    static final Position DEFAULT_POSITION = new Position(ZERO);
```
DEFAULT_POSITION 캐싱을 적용해주셨네요! 👍🏿

yxxnghwan commented on src/main/java/racinggame/domain/RacingGame.java:
```diff
@@ -0,0 +1,35 @@
+        this.moveCondition = moveCondition;
+    }
+
+    public static RacingGame of(Cars cars, MoveCondition moveCondition) {
+
+        return new RacingGame(cars, moveCondition);
+    }
```
그냥 생성자를 호출하는데 정적 메소드로 분리한 이유가 있나요?

kunsanglee replied:
이전 단계에서 리팩터링중에 RacingGame의 of 메서드 안에서 수행하던 검증 로직을 Cars 클래스로 옮기면서
단순히 RacingGame 생성자를 호출하고 반환하는 로직만 남게 되었습니다.

CarInfo, RoundResult와 같은 DTO 성격이 아닌 도메인 클래스에 이미 from, of 와 같은 new 키워드를 통한 생성을 막고 팩토리 메서드를 사용하여 생성하고 있어서 생성시 통일성을 위해 삭제하지 않았는데요
이런 상황에는 과감히 RacingGame 클래스에는 팩토리 메서드를 삭제하는 것이 좋을지 고민이 됩니다!

알렉스는 어떻게 생각하시나요?

yxxnghwan reviewed (approved):
리뷰 반영하느라 수고하셨습니다 이상! 이번 단계에선 충분한 것 같아서 이만 머지하도록 할게요!
2단계도 화이팅입니다! 🚀
